crates/holochain/src/conductor/conductor.rs

```diff
@@ -145,44 +144,103 @@ pub use state_impls::*;
 
 mod graft_records_onto_source_chain;
 
+/// How long we should attempt to achieve a "well-connected" CellStatus when first activating a cell,
+/// before moving on and letting the network health activity go on in the background.
+pub const JOIN_NETWORK_WAITING_PERIOD: std::time::Duration = std::time::Duration::from_secs(5);
+
+
 /// A list of Cells which failed to start, and why
 pub type CellStartupErrors = Vec<(CellId, CellError)>;
 
 /// Cloneable reference to a Conductor
 pub type ConductorHandle = Arc<Conductor>;
 
-/// The reason why a cell is waiting to join the network.
+/// The status of an installed Cell, which captures different phases of its lifecycle
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
-pub enum PendingJoinReason {
-    /// The initial state, no attempt has been made to join the network yet.
-    Initial,
+pub enum CellStatus {
+    /// The cell is fully "online" and well-connected to a good number of peers
+    /// who know of its existence.
+    ///
+    /// This is the desired state for a cell, and should be maintained.
+    WellConnected,
+
+    /// The cell is connected to some peers, but there is evidence that there could
+    /// be more done to get to a better-connected state. Gossip may be weak and
+    /// network performance in general may be poor.
+    PartiallyConnected(String),
+
+    /// The Cell has no peers, or couldn't connect to its peers. Peers don't know about
+    /// this cell, so network activity will likely be impossible. This could be due to
+    /// not being connected to any network at all (being "offline"), or there could be
+    /// no peers to connect to.
+    /// 
+    /// This is also the initial state of a Cell, before it has been fully initialized.
+    /// TODO: perhaps we want to distinguish between these two cases, for different
+    /// retry behavior?
+    ///
+    /// Periodic attempts to discover more peers should be made.
+    Isolated,
 
-    /// The join failed with an error that is safe to retry, such as not being connected to the internet.
-    Retry(String),
+    /// The cell experienced an error which indicates that no more attempts should be made
+    /// to increase network connectivity for this cell. Any app which depends on this cell
+    /// will be put into a Paused state.
+    Unrecoverable(String),
```

Comment:
I'm not sure this should be a state. We can even retry errors like no network interfaces found... some hardware wifi switches actually hotswap the interfaces in and out. If something is truly unrecoverable it seems like it is a bug, and should panic the whole holochain process.